### DISCUSSION TITLE
Build Error: Updated dependencies causing build error

### DISCUSSION THREAD

**Discussion by niallnsec** [projectdiscovery/nuclei#3775] [Q&A]
### Current Behavior:
Nuclei fails to build when used as a library with the following error:
```
# github.com/projectdiscovery/nuclei/v2/pkg/progress
../../../../go/pkg/mod/github.com/projectdiscovery/nuclei/v2@v2.9.6/pkg/progress/progress.go:102:34: undefined: clistats.PrintCallback
# github.com/projectdiscovery/nuclei/v2/pkg/operators/common/dsl
../../../../go/pkg/mod/github.com/projectdiscovery/nuclei/v2@v2.9.6/pkg/operators/common/dsl/dsl.go:85:19: cannot use rawResp.SOA (variable of type []retryabledns.SOA) as []string value in map literal
```

### Expected Behaviour:
Nuclei builds correctly.

### Steps To Reproduce:
In a project using Nuclei as a library, run go get -u and attempt to build.

### Anything else:
This appears to be a problem caused by two dependencies.
1. github.com/projectdiscovery/clistats - Nuclei is using version 0.0.12, the package has since been updated with breaking changes but the semantic versioning guarantees were not upheld. As a result, go get -u updates the package to version 0.0.18, in which the exported type PrintCallback has been removed.
2. github.com/projectdiscovery/retryabledns - Nuclei is using version v1.0.24, similarly to the first issue, a breaking change was introduced in v1.0.26 where the type of the SOA field in the DNS data has been changed form a string array to an array of SOA structures.

As a workaround, anyone using Nuclei as a library can manually pin the version of github.com/projectdiscovery/clistats to v0.0.12 and the version of github.com/projectdiscovery/retryabledns to v1.0.25.

**Comment by MetzinAround** [comment]
Thanks for including a workaround in the issue and the notice about some breaking changes. 

Tagging @ehsandeep to check on whether these have been addressed in the dev branch or not.

**Comment by ehsandeep** [comment]
@niallnsec `go get -u` will upgrade the dependencies, any specific reason for doing it instead of using the locked/compatible version in `go.mod`?

**Comment by Caratpine** [comment]
same issue with me.
Nuclei version: 2.9.6
```
# github.com/projectdiscovery/nuclei/v2/pkg/operators/common/dsl
../../../go/pkg/mod/github.com/projectdiscovery/nuclei/v2@v2.9.6/pkg/operators/common/dsl/dsl.go:85:19: cannot use rawResp.SOA (variable of type []retryabledns.SOA) as []string value in map literal
```

**Comment by Caratpine** [comment]
It was normal when I only use nuclei as library, but when I continued to use katana as library, the above error was reported
Katana version: 1.0.2

**Comment by tarunKoyalwar** [comment]
@Caratpine running 
```go
  go get -u github.com/projectdiscovery/retryabledns@v1.0.24
  go mod tidy
```
should fix the issue. go automatically upgrades minor versions of packages and if there are any breaking changes it cause build failure
downgrading that package manually fixes the issue
if future if you face any similar issue just check if dependencies version in your go.mod file and https://github.com/projectdiscovery/nuclei/blob/main/v2/go.mod are the same . if not downgrade/upgrade them

**Comment by dogancanbakir** [comment]
This discussion closed automatically due to inactivity. Feel free to reopen or start new if still relevant.
